File: core/src/main/java/org/aresclient/ares/api/Ares.java
package org.aresclient.ares.api;

import dev.tigr.simpleevents.EventManager;
import dev.tigr.simpleevents.listener.EventHandler;
import dev.tigr.simpleevents.listener.EventListener;
import kotlin.Unit;
import kotlinx.serialization.json.JsonElement;
import net.meshmc.mesh.loader.MeshLoader;
import net.meshmc.mesh.loader.Mod;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.aresclient.ares.AresStatics;
import org.aresclient.ares.api.command.Command;
import org.aresclient.ares.api.event.AresEvent;
import org.aresclient.ares.api.event.AresEventManager;
import org.aresclient.ares.api.event.client.InputEvent;
import org.aresclient.ares.api.event.client.ShutdownEvent;
import org.aresclient.ares.api.event.client.TickEvent;
import org.aresclient.ares.api.event.render.RenderEvent;
import org.aresclient.ares.api.global.Global;
import org.aresclient.ares.api.minecraft.Minecraft;
import org.aresclient.ares.api.module.Module;
import org.aresclient.ares.api.setting.Setting;
import org.aresclient.ares.impl.JsonSettingSerializer;
import org.aresclient.ares.api.render.MatrixStack;
import org.aresclient.ares.api.render.Renderer;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class Ares {
    public static class Plugin {
        private String id;
        private String name;
        private String description;
        private String version;
        private String[] authors;

        private final List<Global> globals = new ArrayList<>();
        private final List<Module> modules = new ArrayList<>();
        private final List<Command> commands = new ArrayList<>();
        private Setting.Map<?> settings;
        private Logger logger;

        public void init(String id, String name, String description, String version, String[] authors) {
            long start = System.currentTimeMillis();

            this.id = id;
            this.name = name;
            this.description = description;
            this.version = version;
            this.authors = authors;

            settings = PLUGIN_SETTINGS.addMap(id);
            logger = LogManager.getLogger(name);

            PLUGINS.add(this);

            init();

            for(Global global: globals) {
                Ares.getEventManager().register(global);
                Ares.getEventManager().register(global.getClass());
            }

            for(Module module: modules) {
                if(module.isListening()) {
                    Ares.getEventManager().register(module);
                    Ares.getEventManager().register(module.getClass());
                }
                MeshLoader.getInstance().registerInterfaces(module);
            }

            logger.info("Loaded {} globals, {} modules and {} commands in {} milliseconds",
                globals.size(), modules.size(), commands.size(), System.currentTimeMillis() - start);
        }

        public void init(Mod mod) {
            init(mod.getId(), mod.getName(), mod.getDescription(), mod.getVersion(), mod.getAuthors());
        }

        public void init() {
        }

        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public String getDescription() {
            return description;
        }

        public String getVersion() {
            return version;
        }

        public String[] getAuthors() {
            return authors;
        }

        public List<Global> getGlobals() {
            return globals;
        }

        public List<Module> getModules() {
            return modules;
        }

        public List<Command> getCommands() {
            return commands;
        }

        public Setting.Map<?> getSettings() {
            return settings;
        }

        public Logger getLogger() {
            return logger;
        }
    }

    private static final List<Plugin> PLUGINS = new ArrayList<>();
    private static final EventManager EVENT_MANAGER = new AresEventManager();

    private static final File SETTINGS_FILE = new File("ares/config/settings.json");
    private static final JsonSettingSerializer SETTINGS_SERIALIZER = new JsonSettingSerializer(settings -> {
        settings.setPrettyPrint(true);
        return Unit.INSTANCE; // kotlin moment
    });
    private static final Setting.Map<JsonElement> SETTINGS = SETTINGS_SERIALIZER.read(SETTINGS_FILE);
    private static final Setting.Map<?> PLUGIN_SETTINGS = SETTINGS.addMap("Plugins");

    private static Minecraft MINECRAFT = null;
    public static Minecraft getMinecraft() {
        if(MINECRAFT == null) MINECRAFT = AresStatics.getMinecraft();
        return MINECRAFT;
    }

    public static List<Plugin> getPlugins() {
        return PLUGINS;
    }

    public static EventManager getEventManager() {
        return EVENT_MANAGER;
    }

    public static Setting.Map<?> getSettings() {
        return SETTINGS;
    }

    public static void tickClient() {
        for(Plugin plugin: PLUGINS) {
            plugin.getGlobals().forEach(Global::tick);
            plugin.getModules().forEach(Module::tick);
        }
    }

    public static void renderHud(float delta, Renderer.Buffers buffers, MatrixStack matrixStack) {
        for(Plugin plugin: PLUGINS) {
            for(Module module: plugin.getModules()) {
                module.renderHud(delta, buffers, matrixStack);
            }
        }
    }

    public static void renderWorld(float delta, Renderer.Buffers buffers, MatrixStack matrixStack) {
        for(Plugin plugin: PLUGINS) {
            for(Module module: plugin.getModules()) {
                module.renderWorld(delta, buffers, matrixStack);
            }
        }
    }

    public static void tickMotion() {
        for(Plugin plugin: PLUGINS) plugin.getModules().forEach(Module::motion);
    }

    @EventHandler
    private static final EventListener<ShutdownEvent> shutdownListener = new EventListener<>(event -> {
        SETTINGS_FILE.mkdirs();
        SETTINGS_SERIALIZER.write(SETTINGS, SETTINGS_FILE);
    });

    @EventHandler
    private static final EventListener<TickEvent> tickEventListener = new EventListener<>(event -> {
        if(event.getEra() != AresEvent.Era.BEFORE) return;
        if(event.getType() == TickEvent.Type.CLIENT) tickClient();
        else if(event.getType() == TickEvent.Type.MOTION) tickMotion();
    });

    @EventHandler
    private static final EventListener<RenderEvent> renderEventListener = new EventListener<>(event -> {
        if(event.getType() == RenderEvent.Type.HUD) {
            Renderer.State state = Renderer.begin2d();
            renderHud(event.getTickDelta(), state.getBuffers(), state.getMatrixStack());
            Renderer.end(state);
        } else if(event.getType() == RenderEvent.Type.WORLD) {
            Renderer.State state = Renderer.begin3d();
            renderWorld(event.getTickDelta(), state.getBuffers(), state.getMatrixStack());
            Renderer.end(state);
        }
    });

    @EventHandler
    private static final EventListener<InputEvent> inputEventListener = new EventListener<>(event -> {
        int key;
        boolean state;

        if(event instanceof InputEvent.Keyboard.Pressed) {
            key = ((InputEvent.Keyboard.Pressed) event).getKey();
            state = true;
        } else if(event instanceof InputEvent.Keyboard.Released) {
            key = ((InputEvent.Keyboard.Released) event).getKey();
            state = false;
        } else if(event instanceof InputEvent.Mouse.Pressed) {
            key = ((InputEvent.Mouse.Pressed) event).getKey();
            state = true;
        } else if(event instanceof InputEvent.Mouse.Released) {
            key = ((InputEvent.Mouse.Released) event).getKey();
            state = false;
        } else return;

        for(Setting.Bind bind: Setting.Bind.getAll()) {
            if(bind.getValue() != key) continue;
            bind.getCallback().accept(state);
        }
    });

    // TODO: COMMAND LISTENERS

    static {
        getEventManager().register(Ares.class);
    }
}

File: src/main/kotlin/org/aresclient/ares/Ares.kt
package org.aresclient.ares

import dev.tigr.simpleevents.listener.EventHandler
import dev.tigr.simpleevents.listener.EventListener
import net.fabricmc.api.ModInitializer
import net.minecraft.client.gui.screen.ChatScreen
import net.minecraft.client.gui.screen.TitleScreen
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.Identifier
import org.aresclient.ares.api.Plugin
import org.aresclient.ares.api.Wrapper
import org.aresclient.ares.api.events.*
import org.aresclient.ares.api.gui.AresScreen
import org.aresclient.ares.api.instruments.Command
import org.aresclient.ares.api.instruments.Instrument
import org.aresclient.ares.api.instruments.Module
import org.aresclient.ares.api.nrender.drawer.HudDrawer
import org.aresclient.ares.api.nrender.drawer.WorldDrawer
import org.aresclient.ares.api.render.Renderer
import org.aresclient.ares.api.setting.MapSetting
import org.aresclient.ares.api.setting.settings.BindSetting
import org.aresclient.ares.impl.AresPlugin
import org.aresclient.ares.impl.util.ChatUtil
import org.aresclient.ares.impl.util.FriendUtil
import org.slf4j.LoggerFactory
import java.io.File

class Ares: ModInitializer, Wrapper {
	companion object {
		private const val MOD_ID = "ares"
		private const val MOD_NAME = "Ares"
		private val LOGGER = LoggerFactory.getLogger(MOD_NAME)
		private val EVENT_MANAGER = AresEventManager()
		private val PLUGINS = ArrayList<Plugin>()
		private val SETTINGS_FILE = File("ares/config/settings.json")
		private val SETTINGS = MapSetting().also {
			try {
				it.read(SETTINGS_FILE)
			} catch (_: Exception) {
			}
		}
		private val COMMAND_PREFIX_SETTING = SETTINGS.addString("CmdPrefix", "-")
		private val HUD_DRAWER = HudDrawer()
		private val HUD_MATRIX_STACK = MatrixStack()
		private val WORLD_DRAWER = WorldDrawer()

		@JvmStatic fun getModId() = MOD_ID
		@JvmStatic fun getModName() = MOD_NAME
		@JvmStatic fun getLogger() = LOGGER
		@JvmStatic fun getEventManager() = EVENT_MANAGER
		@JvmStatic fun getPlugins() = PLUGINS
		@JvmStatic fun getSettingsFile() = SETTINGS_FILE
		@JvmStatic fun getSettings() = SETTINGS
		@JvmStatic fun getCommandPrefixSetting() = COMMAND_PREFIX_SETTING
		@JvmStatic fun getHudDrawer() = HUD_DRAWER
		@JvmStatic fun getWorldDrawer() = WORLD_DRAWER

		@JvmStatic
		fun identifier(path: String): Identifier = Identifier.of(MOD_ID, path)
	}

	@field:EventHandler
	private val tickEventListener = EventListener<TickEvent> { event ->
		if(event.era != Era.BEFORE) return@EventListener
		else if(event is TickEvent.Client) PLUGINS.forEach(Plugin::tickClient)
		else if(event is TickEvent.Motion) PLUGINS.forEach(Plugin::tickMotion)
	}

	@field:EventHandler
	private val renderEventListener = EventListener<RenderEvent> { event ->
		if(event is RenderEvent.Hud) {
			HUD_DRAWER.begin()
			PLUGINS.forEach { plugin ->
				plugin.renderHud(HUD_DRAWER, HUD_MATRIX_STACK, event.tickDelta)
			}
			HUD_DRAWER.draw()
		} else if(event is RenderEvent.World) {
			WORLD_DRAWER.begin()
			PLUGINS.forEach { plugin ->
				plugin.renderWorld(WORLD_DRAWER, event.tickDelta)
			}
			WORLD_DRAWER.draw()
		}
	}

	@field:EventHandler
	private val inputEventListener = EventListener<InputEvent> { event ->
		if(event.textboxFocused || MC.currentScreen !is TitleScreen && MC.currentScreen !is AresScreen && MC.currentScreen != null)
			return@EventListener

		when(event) {
			is InputEvent.Keyboard.Pressed  -> BindSetting.getAll().forEach { it.triggerCallback(event.key, true, event.repeat) }
			is InputEvent.Keyboard.Released -> BindSetting.getAll().forEach { it.triggerCallback(event.key, false, false) }
			is InputEvent.Mouse.Pressed     -> BindSetting.getAll().forEach { it.triggerCallback(event.key, true, event.repeat) }
			is InputEvent.Mouse.Released    -> BindSetting.getAll().forEach { it.triggerCallback(event.key, false, false) }
		}
	}

	private fun BindSetting.triggerCallback(key: Int, state: Boolean, repeat: Boolean) {
		if(this.value != key) return
		this.callback.accept(state, repeat)
	}

	@field:EventHandler
	private val chatListener = EventListener<ChatEvent> { event ->
		if(event.message.startsWith(COMMAND_PREFIX_SETTING.value)) {
			Command.execute(ChatUtil, event.message.substring(COMMAND_PREFIX_SETTING.value.length))
			MC.inGameHud.chatHud.addToMessageHistory(event.message)
			event.isCancelled = true
		}
	}

	@field:EventHandler
	private val charTypedListener = EventListener<CharTypedEvent> { event ->
		if(MC.currentScreen == null && !MC.NULL && COMMAND_PREFIX_SETTING.value.length == 1 && event.codePoint.toChar() == COMMAND_PREFIX_SETTING.value[0])
			MC.setScreen(ChatScreen(""))
	}

	@field:EventHandler
	private val shutdownListener = EventListener<ShutdownEvent> {
		FriendUtil.save()
		SETTINGS_FILE.parentFile.mkdirs()
		SETTINGS.write(SETTINGS_FILE)
		Renderer.cleanup()
	}

	override fun onInitialize() {
		val start = System.currentTimeMillis()

		EVENT_MANAGER.register(this)

		FriendUtil // TODO: make a global variable

		loadPlugin(AresPlugin)
		// TODO: Dynamic Plugin Loading

		LOGGER.info("Ares loaded {} plugins in {} milliseconds.", PLUGINS.size, System.currentTimeMillis() - start)
	}

	private fun loadPlugin(plugin: Plugin) {
		val start = System.currentTimeMillis()

		plugin.globals.forEach(Instrument::registerEvents)
		plugin.modules.forEach(Instrument::registerEvents)
		plugin.modules.filter(Module::isEnabled).forEach(Module::onEnable)

		plugin.init()
		PLUGINS.add(plugin)

		Companion.LOGGER.info(
			"Loaded plugin {} with {} globals, {} modules and {} commands in {} milliseconds.",
			plugin.name, plugin.globals.size, plugin.modules.size, plugin.commands.size,
			System.currentTimeMillis() - start
		)
	}
}
